### ISSUE TITLE
 Timezone wrong in glib

### ISSUE THREAD

**Issue by Neo-Oli** [termux/termux-packages#448]
Profanity doesn't recognize the timezone and just assumes 0000. 
It uses the following code to determine the timezone:
`tz = g_time_zone_new_local();`

I am trying to test the function on its own, but I will need to wait on #447 (maybe this problem gets also solved by #447)

**Comment by fornwall** [comment]
Hm, can it be due to the `TZ` environment variable not being setup by Android?

It seems like [glib expects this](https://github.com/GNOME/glib/blob/35b401c8bb0c7bb9324e7d8ae0e5d0a63f98872c/glib/gtimezone.c#L1483):

```C
GTimeZone *
g_time_zone_new_local (void)
{
  return g_time_zone_new (getenv ("TZ"));
}
```

Where `g_time_zone_new(NULL)` seems to fallback to `/etc/localtime` (and possibly using `/usr/share/zoneinfo`), which also does not exist on Android.

**Comment by fornwall** [comment]
Hm, the following file seems to be used by bionic: `/system/usr/share/zoneinfo/tzdata`

Also, the following shell call, `getprop persist.sys.timezone`, seems to give me the current timezone. This can also be called in C using:

```C
#include <stdio.h>
#include <sys/system_properties.h>

int main() {
        char timezone_value[PROP_VALUE_MAX];
        if (__system_property_get("persist.sys.timezone", timezone_value) < -1) {
                fprintf(stderr, "__system_property_get(\"persist.sys.timezone\") failed\n");
                return 1;
        }
        printf("timezone = %s\n", timezone_value);
}
```

**Comment by Neo-Oli** [comment]
Shouldn't setting the TZ environment variable manually fix this then? Right now, if I set the TZ with `export TZ=$(getprop persist.sys.timezone)` (getprop [...] uses the same format) it doesn't seem to affect profanity, even though `date` will recognize new timezones this way. 
```
$ export TZ=Asia/Kolkata
$ date
Mon Sep 12 18:01:22 IST 2016
$ export TZ=$(getprop persist.sys.timezone)
$ echo $TZ
Europe/Zurich
$ date
Mon Sep 12 14:31:52 CEST 2016
```

**Comment by fornwall** [comment]
Setting `TZ` to e.g. Europe/Zurich, makes glib look for a `/usr/share/zoneinfo/Europe/Zurich` file, which does not exist on Android.

I guess the first step is to make glib, if `TZ` is not set, default to the `persist.sys.timezone` property.

The next step would be to either:

1. Patch glib to change ``/usr/share/zoneinfo` => `$PREFIX/share/zoneinfo`, and add a `tzdata` package providing that directory.
2. Patch glib to use what the Android system already has for timezone information (`/system/usr/share/zoneinfo/tzdata` perhaps)? I guess this is better, so that glib-using programs on Termux does not differ from the rest of the Android system in timezone handling.

**Comment by fornwall** [comment]
Found http://blog.kir.elagin.me/2014/10/updating-android-cyanogenmod-tzdata.html:

> Unlike Linux, Android has all the timezone database in a single file located at /system/usr/share/zoneinfo/tzdata. The file is binary but it starts with the string tzdata followed by the version in ASCII.

**Comment by fornwall** [comment]
I created an issue at the glib bugzilla: https://bugzilla.gnome.org/show_bug.cgi?id=771304

**Comment by fornwall** [comment]
A small test program:

```C
// Compile with
// clang test.c `pkg-config --cflags --libs glib-2.0` -o test
#include <glib.h>
#include <stdio.h>
#include <stdlib.h>

void test_with_env(char* tzvalue) {
        if (tzvalue == NULL) {
                unsetenv("TZ");
        } else {
                setenv("TZ", tzvalue, 1);
        }
        GTimeZone* tz = g_time_zone_new_local();
        GDateTime* now = g_date_time_new_now(tz);
        gchar* formatted = g_date_time_format(now, "%Y-%m-%e %R");
        char* name = (tzvalue == NULL) ? "" : tzvalue;
        printf("with TZ=%s the current date is: %s\n", name, formatted);
}

int main() {
        test_with_env(NULL);
        test_with_env("Europe/Helsinki");
        test_with_env("US/Eastern");
}
```

**Comment by fornwall** [comment]
Fixed by https://github.com/termux/termux-packages/commit/8aef3a59c9513f83200e575d9bdcb04c3251b268, available in version `2.48.2-2` of the glib package.
